datafusion/physical-plan/src/tree_node.rs

```diff
@@ -39,8 +39,14 @@ impl DynTreeNode for dyn ExecutionPlan {
     }
 }
 
-/// A node object beneficial for writing optimizer rules, encapsulating an [`ExecutionPlan`] node with a payload.
-/// Since there are two ways to access child plans—directly from the plan and through child nodes—it's recommended
+/// A node context object beneficial for writing optimizer rules.
+/// This context encapsulating an [`ExecutionPlan`] node with a payload.
+///
+/// Since each wrapped node has it's children within both the [`PlanContext.plan.children()`],
```

Comment:
👍 

this is indeed quite subtle -- it would be awesome if we could find some way to make it harder to forget to call `update_plan_from_children`